boot/flags.go

```diff
@@ -0,0 +1,251 @@
+	}
+	bl, err := bootloader.Find(rootDir, opts)
+	if err != nil {
+		return err
+	}
+
+	return bl.SetBootVars(map[string]string{
+		"snapd_boot_flags": s,
+	})
+}
+
+// InitramfsActiveBootFlags returns the set of boot flags that are currently set
+// for the current boot, by querying them directly from the source. This method
+// is only meant to be used from the initramfs, since it may query the bootenv
+// or query the modeenv depending on the current mode of the system.
+// For detecting the current set of boot flags outside of the initramfs, use
+// BootFlags(), which will query for the runtime version of the flags in /run
+// that the initramfs will have setup for userspace.
+// Only to be used on UC20+ systems with recovery systems.
+func InitramfsActiveBootFlags(mode string) ([]string, error) {
+	switch mode {
+	case ModeRecover:
+		// no boot flags are consumed / used on recover mode, so return nothing
+		return nil, nil
+
+	case ModeRun:
+		// boot flags come from the modeenv
+		modeenv, err := ReadModeenv(InitramfsWritableDir)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO: consider passing in the modeenv or returning the modeenv here
+		// to reduce the number of times we read the modeenv ?
+		return modeenv.BootFlags, nil
+
+	case ModeInstall:
+		// boot flags always come from the bootenv for the recover bootloader
+		// in install mode
+
+		opts := &bootloader.Options{
+			Role: bootloader.RoleRecovery,
+		}
+		bl, err := bootloader.Find(InitramfsUbuntuSeedDir, opts)
+		if err != nil {
+			return nil, err
+		}
+
+		m, err := bl.GetBootVars("snapd_boot_flags")
+		if err != nil {
+			return nil, err
+		}
+
+		return splitFlagString(m["snapd_boot_flags"]), nil
+
+	default:
+		return nil, fmt.Errorf("internal error: unsupported mode %q", mode)
+	}
+}
+
+// InitramfsSetBootFlags sets the boot flags for the current boot in the /run
+// file that will be consulted in userspace by BootFlags() below. It is meant to
+// be used only from the initramfs.
+func InitramfsSetBootFlags(flags []string) error {
+	// when we are processing boot flags from the initramfs, don't enforce the
+	// allow list such that an old initramfs doesn't drop new boot flags that
+	// userspace snapd understands
+
+	// we do however drop empty boot flags to protect against the fact that
+	// in install mode, the bootenv is untrusted input, so we want some
+	// assurance that garbage is not being propagated through the system
+	// inadvertently
+	nonEmptyFlags := dropEmptyBootFlags(flags)
+	s := serializeFlagString(nonEmptyFlags)
+
+	if err := os.MkdirAll(dirs.SnapBootstrapRunDir, 0755); err != nil {
+		return err
+	}
+
+	return ioutil.WriteFile(snapBootstrapBootFlagsFile, []byte(s), 0644)
+}
+
+// BootFlags returns the current set of boot flags that were active when this
+// device was booted, using a file in /run/ instead of the source of truth. This
+// is to reduce ambiguity about which flags are active for this boot versus the
```

Comment:
this comment about ambiguity seems more something for a code comment somewhere than the doc comment.

Reply:
moved and expanded on in the code comment